pkgs/stdenv/generic/make-derivation.nix

```
@@ -537,8 +537,8 @@ else let
     assert lib.assertMsg (overlappingNames == [ ])
       "The ‘env’ attribute set cannot contain any attributes passed to derivation. The following attributes are overlapping: ${lib.concatStringsSep ", " overlappingNames}";
     lib.mapAttrs
-      (n: v: assert lib.assertMsg (lib.isString v || lib.isBool v || lib.isInt v || lib.isDerivation v)
-        "The ‘env’ attribute set can only contain derivation, string, boolean or integer attributes. The ‘${n}’ attribute is of type ${builtins.typeOf v}."; v)
+      (n: v: assert lib.assertMsg (lib.isString v || lib.isBool v || lib.isInt v || lib.isDerivation v || lib.isList v)
+        "The ‘env’ attribute set can only contain derivation, list, string, boolean or integer attributes. The ‘${n}’ attribute is of type ${builtins.typeOf v}."; (if lib.isList v then toString v else v))
```

I don't think `toString` is the right behavior.
IIUC, `env` is treated as a normal string coercion, and not a `toString` invocation. This affects the list item interpretation of paths:

```
nix-repl> "${./lib}"
"/nix/store/68rmyx4vrvghbkkqlhvb9kv63n761nvc-lib"

nix-repl> "${toString [./lib]}"
"/home/user/nixpkgs/lib"
```

I expect the same inconsistency to occur in `env` (although tbh I didn't check). This is bad. Paths should always behave the same. Their default interpretation is as a vehicle for inserting sources into derivations, and I think we should stick to that. The eval-time location of the sources has nothing to do with the build.

Furthermore, in many, but not all cases, a separator would be expected to be inserted, usually `:`. I don't think we should make any assumptions about this either. Explicit `concatStringsSep` is good for reading and understanding, whereas implicit behavior is where we expose ourselves to silly bugs and overcomplicated compatibility behaviors. There's even a performance cost. `mkDerivation` is basically our most low level function when it comes to packaging, so we better be careful.

---

> Furthermore, in many, but not all cases, a separator would be expected to be inserted

I think it would be fine, and maybe desirable, to resort to the basic `derivation` bare attributes semantics, instead of making new ones here. My opinion is that moving things that should be in the environment and not bash variables should be limited to only moving the attribute from the bare derivation to the `env` attrset. Any different behaviour would be a cause for surprises.

(everything else prior I have no issues with.)

---

Moving something to `env` is a manual action. We may as well use this opportunity to "ask" the author to make their intent explicit. A surprise is ok if it leads to a better outcome overall.

```suggestion
        "The ‘env’ attribute set can only contain derivation, string, boolean or integer attributes. The ‘${n}’ attribute is of type ${builtins.typeOf v}.${optionalString (lib.isList v) " ‘env’ requires you to be explicit about how the list should be rendered. Use for example `lib.concatStringsSep \":\" (<your_value>)` and consider whether escaping is needed."}"; v)
```

I've probably butchered the code with that suggestion, but you catch my drift. I chose `:` because I think that's the most likely candidate in the general case. Moving something to `env` seems a bit odd to me, whereas adding a `:`-separated environment variable seems much more likely.